lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -19,12 +19,47 @@
 #include "mlir/IR/Builders.h"
 #include "llvm/ADT/TypeSwitch.h"
 
-#include "mlir/IR/Verifier.h"
-
 using namespace mlir;
 using namespace circt;
 using namespace pipeline;
 
+namespace {
+static constexpr std::string_view kEnablePortName = "enable";
+static constexpr std::string_view kStallPortName = "stall";
+static constexpr std::string_view kClockPortName = "clk";
+static constexpr std::string_view kResetPortName = "rst";
+static constexpr std::string_view kValidPortName = "valid";
+} // namespace
+
+// Inlines the module pointed to by 'inst' if it is empty. This assumes
+// that 'inst' is the only user of the module. Should probably implement some
+// more generic inlining code for this, but it's simple enough to do when we
+// know that the module is empty.
+static void inlineIfEmpty(hw::InstanceOp inst) {
+  auto mod = cast<hw::HWModuleLike>(inst.getReferencedModule());
+  if (mod->getNumRegions() == 0)
+    return; // Nothing to do.
+
+  Block &body = mod->getRegion(0).front();
+  auto &ops = body.getOperations();
+  if (ops.size() > 1)
+    return; // non-empty
+
+  hw::OutputOp output = cast<hw::OutputOp>(ops.front());
+  DenseMap<Value, Value> valueMapping;
+  for (auto [instOperand, modOperand] :
+       llvm::zip(inst.getOperands(), output.getOperands()))
+    valueMapping[modOperand] = instOperand;
+  llvm::SmallVector<Value> returnValues;
+  for (auto result : output.getOperands())
+    returnValues.push_back(valueMapping[result]);
+
+  // Replace the instance results with the mapped return values.
+  inst.replaceAllUsesWith(returnValues);
+  inst.erase();
+  mod.erase();
```

Comment:
This assumption makes me uncomfortable... Do you mind renaming the function to imply this behavior?

Reply:
Renamed and elaborated the comment.